server/src/main/java/org/elasticsearch/common/xcontent/ChunkedToXContentBuilder.java

```diff
@@ -0,0 +1,279 @@
+        forEach(items, create);
+        endArray();
+        return this;
+    }
+
+    public <T> ChunkedToXContentBuilder array(
+        String name,
+        Iterator<T> items,
+        Function<? super T, CheckedBiConsumer<XContentBuilder, ToXContent.Params, IOException>> create
+    ) {
+        startArray(name);
+        forEach(items, create);
+        endArray();
+        return this;
+    }
+
+    public ChunkedToXContentBuilder field(String name) {
+        addChunk((b, p) -> b.field(name));
+        return this;
+    }
+
+    public ChunkedToXContentBuilder field(String name, long value) {
+        addChunk((b, p) -> b.field(name, value));
+        return this;
+    }
+
+    public ChunkedToXContentBuilder field(String name, String value) {
+        addChunk((b, p) -> b.field(name, value));
+        return this;
+    }
+
+    public ChunkedToXContentBuilder field(String name, Object value) {
+        addChunk((b, p) -> b.field(name, value));
+        return this;
+    }
+
+    /**
+     * Passes this object to {@code consumer}.
+     * <p>
+     * This may seem superfluous, but it allows callers to 'break out' into more declarative code
+     * inside the lambda, whilst maintaining the top-level fluent method calls.
+     */
+    public ChunkedToXContentBuilder execute(Consumer<ChunkedToXContentBuilder> consumer) {
+        consumer.accept(this);
+        return this;
+    }
+
+    public <T> ChunkedToXContentBuilder forEach(Iterator<T> items, BiConsumer<? super T, ChunkedToXContentBuilder> create) {
+        items.forEachRemaining(t -> create.accept(t, this));
+        return this;
+    }
+
+    public <T> ChunkedToXContentBuilder forEach(
+        Iterator<T> items,
+        Function<? super T, CheckedBiConsumer<XContentBuilder, ToXContent.Params, IOException>> create
+    ) {
+        items.forEachRemaining(t -> append(create.apply(t)));
+        return this;
+    }
+
+    public <T> ChunkedToXContentBuilder appendEntries(Map<String, ?> map) {
+        return forEach(map.entrySet().iterator(), (e, b) -> b.field(e.getKey(), e.getValue()));
+    }
+
+    public <T> ChunkedToXContentBuilder appendXContentObjects(Map<String, ? extends ToXContent> map) {
+        return forEach(map.entrySet().iterator(), (e, b) -> b.object(e.getKey(), ob -> ob.appendXContent(e.getValue())));
+    }
+
+    public <T> ChunkedToXContentBuilder appendXContentFields(Map<String, ? extends ToXContent> map) {
+        return forEach(map.entrySet().iterator(), (e, b) -> b.field(e.getKey()).appendXContent(e.getValue()));
+    }
+
+    private ChunkedToXContentBuilder appendXContent(ToXContent xContent) {
+        if (xContent != ToXContent.EMPTY) {
+            addChunk(xContent);
+        }
+        return this;
+    }
+
+    /**
+     * Adds an {@code ToXContent}-like object to the builder.
+     * <p>
+     * This uses a straight {@code BiConsumer} rather than {@code ToXContent} here
+     * so the lambda is not forced to return the builder back to us.
```

Comment:
While I do agree that this is sometimes neater, it's kinda weird to have this ability just here. I'd rather we stuck with the proper `XContent` interface that returns the builder, and if we really want to stop returning the builder then we can do it in a more wholesale way.

Reply:
OK - I'll leave that to another time